btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/WithdrawalConsensusProvider.kt

```diff
@@ -5,204 +5,162 @@
+    private val btcNetworkConfigProvider: BtcNetworkConfigProvider
 ) {
-    private val gson = GsonInstance.get()
 
     /**
      * Creates consensus data and saves it in Iroha
      * @param withdrawalDetails - withdrawal details that will be used to create consensus
      */
     fun createConsensusData(withdrawalDetails: WithdrawalDetails): Result<Unit, Exception> {
-        //Create consensus storage for withdrawal
-        val consensusAccountName = withdrawalDetails.irohaFriendlyHashCode()
-        val consensusAccountId = "$consensusAccountName@$BTC_CONSENSUS_DOMAIN"
-        return bitcoinUTXOProvider.getAvailableUTXOHeight(
-            withdrawalDetails,
-            bitcoinConfig.confidenceLevel,
-            withdrawalDetails.withdrawalTime
-        ).fanout {
-            withdrawalQueryHelper.getPeersCount()
-        }.map { (availableHeight, peers) ->
-            val withdrawalConsensus =
-                WithdrawalConsensus(
-                    availableHeight,
-                    peers
-                )
-            consensusIrohaConsumer.send(
-                IrohaConverter.convert(
-                    createAccountTx(
-                        consensusAccountName,
-                        withdrawalDetails
-                    )
-                )
-            ).failure { ex ->
-                throw D3ErrorException.fatal(
-                    failedOperation = WITHDRAWAL_OPERATION,
-                    description = "Cannot create consensus data storage account for withdrawal $withdrawalDetails",
-                    errorCause = ex
-                )
+        return hasBeenEstablished(withdrawalDetails.irohaFriendlyHashCode()).flatMap { hasBeenEstablished ->
+            // No need to create consensus if it has been established before
+            if (hasBeenEstablished) {
+                return@flatMap Result.of(Unit)
             }
-            withdrawalConsensus
-        }.map { withdrawalConsensus ->
-            consensusIrohaConsumer.send(
-                IrohaConverter.convert(addConsensusDataTx(consensusAccountId, withdrawalConsensus, withdrawalDetails))
-            ).fold(
+            handleConsensus(withdrawalDetails)
+        }
+    }
+
+    /**
+     * Handle consensus creation process
+     * @param withdrawalDetails - details of withdrawal
+     * @return result of operation
+     */
+    private fun handleConsensus(withdrawalDetails: WithdrawalDetails): Result<Unit, Exception> {
+        val utxo = ArrayList<SerializableUTXO>()
+        val consensusId = UUID.randomUUID().toString()
+        // Collect unspents
+        return bitcoinUTXOProvider.collectUnspents(withdrawalDetails, bitcoinConfig.confidenceLevel)
+            .flatMap { unspents ->
+                unspents.forEach { output ->
+                    val transaction = Transaction(btcNetworkConfigProvider.getConfig())
+                    val input = transaction.addInput(output)
+                    input.setParent(null)
```

Comment:
just to reduce memory consumption or what?

Reply:
No. It's a trick. The constructor I want to use is not public and the only way to invoke it is by using `addInput` which returns an input. Unfortunately,  `addInput` has side effects. For instance, it sets `transaction` as an input parent transaction which is not needed. This is why I set it to null.